### DISCUSSION TITLE
Registration Space of "Rigid-body" Registration

### DISCUSSION THREAD

**Discussion by root-at-rest** [SuperElastix/elastix#877] [Q&A]
Hello

I'm working on a comparison study on Registration algorithms in Case of "Rigid-body" Registration of 3D T1-weighted MRI Structural images and elastix is just working fantastic and fast so i wanted to include your amazing job in my comparison study.

i need help with something like Orientations, Parameters Interpretation, ... and I am seeking your advice
 
my concerns are:

1. finding the **4*4** affine matrices from rigid-body registration output file (as FreeSurfer and FSL does): 
   from  **elastix's "Rigid-body"** output file (6 numbers/parameters) i wanted to findout **the  4*4 registration matrix** from them to compare them with other programs like FSL and FreeSurfer.

2. Does elastix uses **the same Orientation** as FSL, Freesurfer considerations generating the Registration matrix ? (is there any flag to make that output which i missed?)

3. if there is a way to find fsl-liked matrices outputted by elastix (by which i mean the matrices that can be applied using fsl as FreeSurfer do so for FSL) I'll be appreciated it if you guide me to do so.

hope this massage find you soon.

Regards
root@rest

**Comment by ntatsisk** [comment]
Hi @root-at-rest,

Some initial remarks:

1. For a 3D affine registration I would expect 12 output parameters (not 6). They can be used to populate the 4x4 matrix in a *row-first* order (3 first rows x 4 parameters = 12 parameters) - the last row contains zeros and a single 1.
2. There is no flag to change the underlying orientation of elastix to match other software. Elatstix is based on ITK which uses an LPS coordinate system (explanation of LPS: https://www.slicer.org/wiki/Coordinate_systems). From a quick google search, it looks like Freesurfer/FSL use RAS (https://surfer.nmr.mgh.harvard.edu/fswiki/CoordinateSystems). 
3. If FSL follows indeed the RAS convention, then the 4x4 matrix needs to be converted from LPS->RAS. We developed originally LPS->pixel space here https://github.com/InsightSoftwareConsortium/itk-torch-transform-bridge/pull/6 but I think somewhere there in the long discussion there is also the LPS->RAS conversion.

Hope this helps for now!

**Reply by root-at-rest** [reply]
Hi @ntatsisk and thanks for your  kind and useful response

I just have a little problem with number 1, 

As i think every Rigid-body transform could be represented in a **4*4 Matrix** (a **3*3 matrix** which is a member of Special Orthogonal SO(3) encoding the Rotation in 3D + a vector in R(3) which encodes the Translation in 3D, and the Last Row as you said so is the [0 0 0 1]  )
FSL, Fresssurfer will output **the 4*4 matrix** (_in case we constrained the degree of freedom DOF to 6 Which means we are computing the Rigid-Body transform_ ) but as i run Elastix with [Parameters_Rigid.txt](https://github.com/SuperElastix/ElastixModelZoo/blob/master/models/default/Parameters_Rigid.txt) i receive a output file as (TransformParameters 0.006598 -0.005876 0.002128 -0.139096 -0.108889 0.106695) , I want to Construct **the Correspondence 4*4 Matrix** from these Rigid-body parameters

**Reply by root-at-rest** [reply]
@ntatsisk
Hi, wish this lines find you well
it's been a while and as i got no answer i hope that replying with a tag would do the followup massage for me as i didn't know any other way, sending a message to you.

Thanks 
@root-at-rest

**Reply by ntatsisk** [reply]
Hi @root-at-rest, tagging me is completely fine! We try our best to respond to questions but of course when other community members chime in to help is always more than welcome. :)

Now back to your question, I don't know answer from the top of my head either, but I will try to share further hints on how I would attempt to find it:
* In my original reply I mistakenly responded for the Affine transform case while you clearly were asking for the Rigid transform - sorry for that. From the [manual](https://elastix.lumc.nl/doxygen/index.html) on _section 2.6_ I read that for a 3D rigid transform that output transform parameters are 

> In 3D, this gives a vector of length 6: μ = (θx, θy , θz , tx, ty , tz )^T 

where the first three parameters are rotations and three last are translations

* So, I would try to create the three corresponding rotation matrices from the rotation parameters (see: https://en.wikipedia.org/wiki/Rotation_matrix#Basic_rotations) and combine them into a single affine matrix using matrix multiplication - the order matters though!
* You can fetch the parameters from the registration result as here: https://github.com/SuperElastix/elastix/discussions/883#discussioncomment-5905907. Do you happen to use python?
* Now assuming that you managed to compute a 3x3 matrix and 3-element translation vector, you can use the ITK resample image filter to resample the moving image and make sure it matches the resulting image from the registration. This step is to double check that you combined the rotations correctly. Examples of `itk.resample_image_filter` (https://github.com/InsightSoftwareConsortium/ITKElastix/blob/main/examples/ITK_Example13_ITKResampling.ipynb and https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_itk_torch_bridge.py#L112-L137)
* There is also this older discussion in case it gives any extra ideas: https://groups.google.com/g/elastix-imageregistration/c/FJy1kkAXgGY/m/g-2DKdfhBwAJ

Let me know how it goes! I can also do some hands on work if you get stuck again. When we get this right, I plan to add a [jupyter notebook](https://github.com/InsightSoftwareConsortium/ITKElastix/tree/main/examples) for future users that might find it useful. So, don't forget to share your code ;)

**Reply by root-at-rest** [reply]
Hi @ntatsisk ,


I have Search through a lot these days to find the appropriate flag which i should add in "Parameter File" so i could use the Command Line(CL) Version not the Python.( I want to Compare the run time of different Apps and this is my reason to run elastix in CL as FSL, FreeSurfer, ants,and atra)  on the Other hand as I am trying to Judge the resultant Matrices I prefer not to interfere Calculations and let Programs do it  
But i wasn't able to navigate a clean and clear site for getting more familiar with Parameter File flags and as Im not used to ITK-based C programming some clear things for an old ITK user seems Completely Complex to me.

Let me Say What I've Found:

An Other ITK-based Registration Tool named ANTS has sided programs(called ConvertTransformFile) which Convert a Registration Parameter to The 4-by-4 Registration Matrix, Considering the "Rotation, Translation, Center, Spacing("Voxel Spacing") representing Mapping of Voxeles to [homogeneous coordinate](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiCstq7j_6AAxWlg_0HHeuNCm4QFnoECCEQAw&url=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FHomogeneous_coordinates&usg=AOvVaw1fQUw9EDJ_UlrpHZEO_jB9&opi=89978449) 
 I think that for Now there might be flags that do such things for elastix as i ran into **WriteITKCompositeTransform "true" ** and i thought that there are something like these flags in Elastix to Consider the whole Params (Rotation, Translation, Center, Spacing) and make the corespondent result.   
Q:  Where Could i find Such Flags for Parameter File (I saw the [doxygen](https://elastix.lumc.nl/doxygen/parameter.html), but i Couldnt find them)

in the Latest Version of elastix it is said that there are new Options like ** WriteITKCompositeTransform ** i thinck there might be Other flags that could help me find the final Registration Matrix ....



Somehow my problem relates to [this](https://github.com/InsightSoftwareConsortium/ITKElastix/issues/79) too.

as you maybe more familiar with The ITK-Core and The elastix Program in General i would be appreciate it if you could help me find Something about probable Flags in the Parameter File.


Bests
@root-at-rest
